### PR TITLE
Cleanup XVal

### PR DESCRIPTION
**Pull request by ptschip**
Encapsulate XVal data in CBlock.  Gets rid of global data, global critsec and all the necessity of cleaning up data here and there.

### CODE REVIEW

File: src/validation/validation.cpp [hunk 1]
```diff
@@ -2321,21 +2306,15 @@ bool ConnectBlockCanonicalOrdering(const CBlock &block,
...
                 nFees += view.GetValueIn(tx) - tx.GetValueOut();
 
-                // Only check inputs when the tx hash in not in the setPreVerifiedTxHash as would only
-                // happen if this were a regular block or when a tx is found within the returning XThinblock.
                 uint256 hash = tx.GetHash();
                 {
+                    // If XVal is not on then check all inputs, otherwise only check
+                    // transactions that were not previously verified in the mempool.
+                    bool fUnVerified = block.setUnVerifiedTxns.count(hash);
+                    if ((block.fXVal && fUnVerified) || !block.fXVal)
```
Comment (sickpig) [inline src/validation/validation.cpp]: Isn't this equivalent to:

```c++
if (!block.fXVal || fUnVerified)
```

**Comment by ptschip** [comment]
yes I suppose you're right, will change...thanks.

the second On 07/02/2019 4:44 a.m., Andrea Suisani wrote:

**Review by sickpig — APPROVED** [review]

File: src/blockrelay/compactblock.cpp [hunk 2]
```diff
@@ -666,12 +656,13 @@ static bool ReconstructBlock(CNode *pfrom, const bool fXVal, int &missingCount,
...
                 ptx = orphanpool.mapOrphanTransactions[hash].ptx;
-                setUnVerifiedOrphanTxHash.insert(hash);
+                pfrom->compactBlock.setUnVerifiedTxns.insert(hash);
             }
-            else if ((inMemPool || inCommitQ) && fXVal)
-                setPreVerifiedTxHash.insert(hash);
             else if (inMissingTx)
+            {
                 ptx = pfrom->mapMissingTx[nShortId];
+                pfrom->compactBlock.setUnVerifiedTxns.insert(hash);
```
Comment (gandrewstone) [inline src/blockrelay/compactblock.cpp]: pfrom->compactBlock.setUnVerifiedTxns.insert(hash); exists in both branches

File: src/blockrelay/graphene.cpp [hunk 3]
```diff
@@ -746,16 +728,19 @@ static bool ReconstructBlock(CNode *pfrom, const bool fXVal, int &missingCount,
...
                 ptx = orphanpool.mapOrphanTransactions[hash].ptx;
-                setUnVerifiedOrphanTxHash.insert(hash);
+                pfrom->grapheneBlock.setUnVerifiedTxns.insert(hash);
             }
-            else if ((inMemPool || inCommitQ) && fXVal)
-                setPreVerifiedTxHash.insert(hash);
             else if (inMissingTx)
+            {
                 ptx = pfrom->mapMissingTx[hash.GetCheapHash()];
+                pfrom->grapheneBlock.setUnVerifiedTxns.insert(hash);
```
Comment (gandrewstone) [inline src/blockrelay/graphene.cpp]: ditto, insert is in both branches

**Pull request merged**